File: Dokie/Navigation/NavigationRouter.m
//
//  NavigationRouter.m
//  Doci
//
//  Created by Ashamaz Shidov on 02/12/2018.
//  Copyright © 2018 Ashamaz Shidov. All rights reserved.
//

#import "NavigationRouter.h"
#import "NavigationRouterInclude.h"
#import "CoreDataManager.h"

#import <UIKit/UIKit.h>

@interface NavigationRouter ()

@property (nonatomic, strong) UINavigationController *navController;
@property (nonatomic, strong) UIWindow *mainWindow;
@property (nonatomic, strong) UIView *loadingView;
@property (nonatomic, strong) NSMutableArray<NSString *> *loadingGuids;
@property (nonatomic, strong) UIBlurEffect *blurEffect;
@property (nonatomic, strong) UIVisualEffectView *blurView;

@property (nonatomic, strong, readonly) CoreDataManager *coreDataManager;

@end

@implementation NavigationRouter

+ (instancetype)shared {
    static id sharedObject = nil;
    static dispatch_once_t onceToken = 0;
    dispatch_once(&onceToken, ^{
        sharedObject = [NavigationRouter new];
    });
    return sharedObject;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        [self setup];
    }
    return self;
}

- (void)setup {
    self.blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleDark];
    self.blurView = [[UIVisualEffectView alloc] initWithEffect:self.blurEffect];
    
    self.loadingGuids = [NSMutableArray array];
}

- (void)createLoginViewController {
    LoginViewModel *viewModel = [LoginViewModel new];
    LoginViewController *rootViewController = [[LoginViewController alloc] initWithViewModel:viewModel];
    
    self.navController = [[UINavigationController alloc] initWithRootViewController:rootViewController];
    self.navController.navigationBar.tintColor = [UIColor blueColor];

    self.mainWindow = [[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
    [self.mainWindow setRootViewController:self.navController];
    [self.mainWindow makeKeyAndVisible];
}

- (void)pushMainMenuWithEncryptionManager:(EncryptionManager *)encryptionManager managedObjectContext:(NSManagedObjectContext *)objectContext {
    [self.coreDataManager setupWithEncryptionManager:encryptionManager managedObjectContext:objectContext];
    [self pushFolder:nil];
}

- (void)pushFolder:(CDFolder * _Nullable)folder {
    FolderViewModel *viewModel = [[FolderViewModel alloc] initWithFolder:folder];
    FolderViewController *viewController = [[FolderViewController alloc] initWithViewModel:viewModel];
    [self.navController pushViewController:viewController animated:YES];
}

- (void)pushDocument:(CDDocument *)document {
    DocumentViewModel *viewModel = [[DocumentViewModel alloc] initWithDocument:document];
    DocumentViewController *viewController = [[DocumentViewController alloc] initWithViewModel:viewModel];
    [self.navController pushViewController:viewController animated:YES];
}

- (void)showAlertWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title
                                                                   message:message
                                                            preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"OK", @"Ok button text in alert view") style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
    [self showAlert:alert];
}

- (void)showAlert:(UIAlertController *)alert {
    UIViewController *on = self.navController.viewControllers.lastObject;
    alert.popoverPresentationController.sourceView = on.view;
    alert.popoverPresentationController.sourceRect = CGRectMake(on.view.bounds.size.width/2, on.view.bounds.size.height/2, 0, 0);
    alert.popoverPresentationController.permittedArrowDirections = 0;
    [self.navController presentViewController:alert animated:YES completion:nil];
}

- (void)showImagePicker:(UIImagePickerController *)imagePicker {
    [self.navController presentViewController:imagePicker animated:YES completion:nil];
}

- (void)shareItems:(NSArray *)items {
    UIActivityViewController * avc = [[UIActivityViewController alloc] initWithActivityItems:items applicationActivities:nil];
    [self.navController presentViewController:avc animated:YES completion:nil];
}

- (void)showDocumentImages:(NSArray<CDContent *> *)contentImages firstIndex:(NSInteger)index {
    DocumentImagesPageViewModel *viewModel = [[DocumentImagesPageViewModel alloc] initWithContentImages:contentImages firstIndex:index];
    DocumentImagesPageViewController *vc = [[DocumentImagesPageViewController alloc] initWithViewModel:viewModel];
    
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:vc];
    [navController setModalPresentationStyle:UIModalPresentationPageSheet];

    [self.navController presentViewController:navController animated:YES completion:nil];
}

- (void)logout {
    [self.coreDataManager reset];
    [self.navController popToRootViewControllerAnimated:YES];
}

- (void)showChangePassword {
    ChangePasswordViewModel *viewModel = [ChangePasswordViewModel new];
    ChangePasswordViewController *vc = [[ChangePasswordViewController alloc] initWithViewModel:viewModel];

    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:vc];
    [navController setModalPresentationStyle:UIModalPresentationPageSheet];

    [self.navController presentViewController:navController animated:YES completion:nil];
}

- (NSString *)showLoading {
    if (!self.loadingView) {
        self.loadingView = [UIView new];
        self.loadingView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.7];
        self.loadingView.translatesAutoresizingMaskIntoConstraints = NO;
        {
            UIActivityIndicatorView *activity = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
            activity.translatesAutoresizingMaskIntoConstraints = NO;
            [activity startAnimating];
            [self.loadingView addSubview:activity];
            [self.loadingView addConstraint:[NSLayoutConstraint constraintWithItem:activity attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self.loadingView attribute:NSLayoutAttributeCenterX multiplier:1 constant:0]];
            [self.loadingView addConstraint:[NSLayoutConstraint constraintWithItem:activity attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self.loadingView attribute:NSLayoutAttributeCenterY multiplier:1 constant:0]];
        }
        NSDictionary *views = NSDictionaryOfVariableBindings(_loadingView);
        [self.navController.view addSubview:self.loadingView];
        [self.navController.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[_loadingView]|" options:0 metrics:nil views:views]];
        [self.navController.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[_loadingView]|" options:0 metrics:nil views:views]];
        [self.navController.view layoutIfNeeded];
    }
    NSString *guid = [NSUUID UUID].UUIDString;
    [self.loadingGuids addObject:guid];
    return guid;
}

- (void)hideLoading:(NSString *)guid {
    [self.loadingGuids removeObject:guid];
    if (self.loadingGuids.count == 0) {
        [self.loadingView removeFromSuperview];
        self.loadingView = nil;
    }
}

- (void)blur {
    self.blurView.frame = self.mainWindow.frame;
    [self.mainWindow addSubview:self.blurView];
}

- (void)unblur {
    [self.blurView removeFromSuperview];
}

- (void)showAbout {
    AboutViewController *vc = [[AboutViewController alloc] init];

    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:vc];
    [navController setModalPresentationStyle:UIModalPresentationPageSheet];

    [self.navController presentViewController:navController animated:YES completion:nil];
}

#pragma mark -

- (CoreDataManager *)coreDataManager {
    return [CoreDataManager shared];
}

@end

File: Dokie/Navigatoin/NavigationRouter.swift
//
//  NavigationRouter.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 14.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import Foundation
import UIKit
import CoreData
import ImagePicker

class NavigationRouter {
    private var navController: UINavigationController!
    private var mainWindow: UIWindow!
    private var loadingView: UIView? = nil
    private var loadingGuids = [String]()
    private let coreDataManager = CoreDataManager.shared

    static let shared = NavigationRouter()
    
    func createLoginViewController() {
        let viewModel = LoginViewModel()
        let viewController = LoginViewController(viewModel: viewModel)
        navController = UINavigationController(rootViewController: viewController)

        mainWindow = UIWindow(frame: UIScreen.main.bounds)
        mainWindow.rootViewController = navController
        mainWindow.makeKeyAndVisible()
    }

    
    func pushMainMenu(encryptionManager: EncryptionManager, managedObjectContext: NSManagedObjectContext) {
        coreDataManager.setup(encryptionManager: encryptionManager, context: managedObjectContext)
        push(folder: nil)
    }
    
    func push(folder: CDFolder?) {
        let viewModel = FolderViewModel(folder: folder)
        let viewController = FolderViewController(viewModel: viewModel)
        navController.pushViewController(viewController)
    }

    func push(document: CDDocument) {
        let viewModel = DocumentViewModel(document: document)
        let viewController = DocumentViewController(viewModel: viewModel)
        navController.pushViewController(viewController)
    }
    
    func showAlert(title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let cancel = UIAlertAction(title: String.localized("OK"), style: .cancel, handler: nil)
        alert.addAction(cancel)
        showAlert(alert)
    }
    
    func showAlert(_ alert: UIAlertController) {
        guard let on = navController.viewControllers.last else { return }
        alert.popoverPresentationController?.sourceView = on.view
        alert.popoverPresentationController?.sourceRect = CGRect(x: on.view.bounds.size.width/2, y: on.view.bounds.size.height/2, width: 0, height: 0)
        alert.popoverPresentationController?.permittedArrowDirections = UIPopoverArrowDirection(rawValue: 0)
        navController.present(alert)
    }
    
    func showImagePicker(_ imagePicker: UIImagePickerController) {
        navController.present(imagePicker)
    }

    func share(items: [Any]) {
        let avc = UIActivityViewController(activityItems: items, applicationActivities: nil)
        navController.present(avc)
    }
    
    func show(images: [CDContent], firstIndex index: Int = 0) {
        let viewModel = DocumentImagesPageViewModel(images: images, firstIndex: index)
        let viewController = DocumentImagesPageViewController(viewModel: viewModel)
        
        let newNavController = UINavigationController(rootViewController: viewController)
        newNavController.modalPresentationStyle = .pageSheet
        navController.present(newNavController)
    }
    
    func logout() {
        coreDataManager.reset()
        navController.popViewController(animated: true)
    }
    
    func showChangePassword() {
        let viewModel = ChangePasswordViewModel()
        let viewController = ChangePasswordViewController(viewModel: viewModel)
        
        let newNavController = UINavigationController(rootViewController: viewController)
        newNavController.modalPresentationStyle = .pageSheet
        navController.present(newNavController)
    }

    func showLoading() -> String {
        if loadingView == nil {
            let view = UIView()
            view.backgroundColor = UIColor.black.withAlphaComponent(0.7)
            view.translatesAutoresizingMaskIntoConstraints = false
            
            let activity: UIActivityIndicatorView
            if #available(iOS 13, *) {
                activity = UIActivityIndicatorView(style: .large)
            }
            else {
                activity = UIActivityIndicatorView(style: .whiteLarge)
            }
            activity.translatesAutoresizingMaskIntoConstraints = false
            activity.startAnimating()
            view.addSubview(activity)
            
            NSLayoutConstraint.activate([activity.centerXAnchor.constraint(equalTo: view.centerXAnchor),
                                         activity.centerYAnchor.constraint(equalTo: view.centerYAnchor)])
            
            let views = ["view" : view]
            navController.view.addSubview(view)
            navController.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[view]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: nil, views: views))
            navController.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[view]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: nil, views: views))
            
            loadingView = view
        }
        
        let guid = UUID().uuidString
        loadingGuids.append(guid)
        return guid
    }

    func hideLoading(_ guid: String) {
        if let index = loadingGuids.firstIndex(of: guid) {
            loadingGuids.remove(at: index)
        }

        if loadingGuids.count == 0 {
            loadingView?.removeFromSuperview()
            loadingView = nil
        }
    }

    func showAbout() {
        let viewController = AboutViewController()

        let newNavController = UINavigationController(rootViewController: viewController)
        newNavController.modalPresentationStyle = .pageSheet
        navController.present(newNavController)
    }

    func showImagePicker(_ picker: ImagePickerController) {
        navController.present(picker)
    }
    
}

extension UINavigationController {
    
    func pushViewController(_ viewController: UIViewController) {
        pushViewController(viewController, animated: true)
    }
    
    func present(_ viewControllerToPresent: UIViewController) {
        present(viewControllerToPresent, animated: true, completion: nil)
    }
    
}
